File: packages/sdks/client-sdks/gatrix-js-client-sdk/tests/FeaturesClient.test.ts
import { FeaturesClient } from '../src/FeaturesClient';
import { EventEmitter } from '../src/EventEmitter';
import { EvaluatedFlag, GatrixClientConfig } from '../src/types';
import { EVENTS } from '../src/events';

// Mock fetch
const mockFetch = jest.fn();
global.fetch = mockFetch;

// Bootstrap data with various flag configurations
const bootstrapFlags: EvaluatedFlag[] = [
    {
        name: 'bool-enabled',
        enabled: true,
        variant: { name: 'default', enabled: true },
        variantType: 'none',
        version: 1,
        reason: 'targeting_match',
    },
    {
        name: 'bool-disabled',
        enabled: false,
        variant: { name: 'disabled', enabled: false },
        variantType: 'none',
        version: 2,
        reason: 'disabled',
    },
    {
        name: 'string-variant',
        enabled: true,
        variant: { name: 'variantA', enabled: true, payload: 'hello world' },
        variantType: 'string',
        version: 3,
        reason: 'percentage_rollout',
    },
    {
        name: 'number-variant',
        enabled: true,
        variant: { name: 'numVariant', enabled: true, payload: '42.5' },
        variantType: 'number',
        version: 4,
        reason: 'evaluated',
    },
    {
        name: 'json-variant',
        enabled: true,
        variant: {
            name: 'jsonVariant',
            enabled: true,
            payload: '{"key":"value","nested":{"a":1}}',
        },
        variantType: 'json',
        version: 5,
        reason: 'evaluated',
        impressionData: true,
    },
    {
        name: 'invalid-number',
        enabled: true,
        variant: { name: 'badNum', enabled: true, payload: 'not-a-number' },
        variantType: 'number',
        version: 6,
    },
    {
        name: 'invalid-json',
        enabled: true,
        variant: { name: 'badJson', enabled: true, payload: 'not valid json' },
        variantType: 'json',
        version: 7,
    },
    {
        name: 'disabled-with-payload',
        enabled: false,
        variant: { name: 'disabled', enabled: false, payload: 'ignored' },
        variantType: 'string',
        version: 8,
        reason: 'disabled',
    },
];

describe('FeaturesClient', () => {
    let emitter: EventEmitter;
    let client: FeaturesClient;
    const defaultConfig: GatrixClientConfig = {
        url: 'https://api.example.com/api/v1/client/features/test/eval',
        apiKey: 'test-api-key',
        appName: 'test-app',
        features: {
            bootstrap: bootstrapFlags,
            disableRefresh: true, // Disable polling for tests
        },
    };

    beforeEach(() => {
        emitter = new EventEmitter();
        client = new FeaturesClient(emitter, defaultConfig);
        mockFetch.mockReset();
    });

    describe('initialization with bootstrap', () => {
        it('should be ready immediately with bootstrap data', async () => {
            await client.init();

            expect(client.isReady()).toBe(true);
        });

        it('should emit READY event', async () => {
            const readyCallback = jest.fn();
            emitter.on(EVENTS.READY, readyCallback);

            await client.init();

            expect(readyCallback).toHaveBeenCalled();
        });

        it('should have all bootstrap flags', async () => {
            await client.init();

            const allFlags = client.getAllFlags();
            expect(allFlags.length).toBe(bootstrapFlags.length);
        });
    });

    describe('isEnabled', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return true for enabled flag', () => {
            expect(client.isEnabled('bool-enabled')).toBe(true);
        });

        it('should return false for disabled flag', () => {
            expect(client.isEnabled('bool-disabled')).toBe(false);
        });

        it('should return false for non-existent flag', () => {
            expect(client.isEnabled('non-existent')).toBe(false);
        });
    });

    describe('boolVariation', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return true for enabled flag', () => {
            expect(client.boolVariation('bool-enabled', false)).toBe(true);
        });

        it('should return false for disabled flag', () => {
            expect(client.boolVariation('bool-disabled', true)).toBe(false);
        });

        it('should return default for non-existent flag', () => {
            expect(client.boolVariation('non-existent', true)).toBe(true);
            expect(client.boolVariation('non-existent', false)).toBe(false);
        });
    });

    describe('stringVariation', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return payload for enabled string flag', () => {
            expect(client.stringVariation('string-variant', 'default')).toBe('hello world');
        });

        it('should return default for disabled flag', () => {
            expect(client.stringVariation('disabled-with-payload', 'default')).toBe('default');
        });

        it('should return default for non-existent flag', () => {
            expect(client.stringVariation('non-existent', 'default')).toBe('default');
        });
    });

    describe('numberVariation', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return parsed number for number flag', () => {
            expect(client.numberVariation('number-variant', 0)).toBe(42.5);
        });

        it('should return default for invalid number', () => {
            expect(client.numberVariation('invalid-number', 999)).toBe(999);
        });

        it('should return default for non-existent flag', () => {
            expect(client.numberVariation('non-existent', 100)).toBe(100);
        });
    });

    describe('jsonVariation', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return parsed JSON for json flag', () => {
            expect(client.jsonVariation('json-variant', {})).toEqual({
                key: 'value',
                nested: { a: 1 },
            });
        });

        it('should return default for invalid json', () => {
            expect(client.jsonVariation('invalid-json', { fallback: true })).toEqual({
                fallback: true,
            });
        });

        it('should return default for non-existent flag', () => {
            expect(client.jsonVariation('non-existent', { default: true })).toEqual({
                default: true,
            });
        });
    });

    describe('variationDetails', () => {
        beforeEach(async () => {
            await client.init();
        });

        describe('boolVariationDetails', () => {
            it('should return correct details for existing flag', () => {
                const result = client.boolVariationDetails('bool-enabled');

                expect(result).toEqual({
                    value: true,
                    reason: 'targeting_match',
                    flagExists: true,
                    enabled: true,
                });
            });

            it('should return flag_not_found for non-existent flag', () => {
                const result = client.boolVariationDetails('non-existent');

                expect(result).toEqual({
                    value: false,
                    reason: 'flag_not_found',
                    flagExists: false,
                    enabled: false,
                });
            });
        });

        describe('stringVariationDetails', () => {
            it('should return correct details for string flag', () => {
                const result = client.stringVariationDetails('string-variant');

                expect(result).toEqual({
                    value: 'hello world',
                    reason: 'percentage_rollout',
                    flagExists: true,
                    enabled: true,
                });
            });

            it('should return disabled reason for disabled flag', () => {
                const result = client.stringVariationDetails('disabled-with-payload', 'default');

                expect(result.reason).toBe('disabled');
                expect(result.enabled).toBe(false);
                expect(result.value).toBe('default');
            });
        });

        describe('numberVariationDetails', () => {
            it('should return correct details for number flag', () => {
                const result = client.numberVariationDetails('number-variant');

                expect(result).toEqual({
                    value: 42.5,
                    reason: 'evaluated',
                    flagExists: true,
                    enabled: true,
                });
            });

            it('should return parse_error for invalid number', () => {
                const result = client.numberVariationDetails('invalid-number', 0);

                expect(result.reason).toBe('parse_error');
                expect(result.value).toBe(0);
            });
        });

        describe('jsonVariationDetails', () => {
            it('should return correct details for json flag', () => {
                const result = client.jsonVariationDetails('json-variant', {});

                expect(result).toEqual({
                    value: { key: 'value', nested: { a: 1 } },
                    reason: 'evaluated',
                    flagExists: true,
                    enabled: true,
                });
            });

            it('should return parse_error for invalid json', () => {
                const result = client.jsonVariationDetails('invalid-json', {});

                expect(result.reason).toBe('parse_error');
            });
        });
    });

    describe('variationOrThrow', () => {
        beforeEach(async () => {
            await client.init();
        });

        describe('boolVariationOrThrow', () => {
            it('should return value for existing flag', () => {
                expect(client.boolVariationOrThrow('bool-enabled')).toBe(true);
            });

            it('should throw for non-existent flag', () => {
                expect(() => client.boolVariationOrThrow('non-existent')).toThrow(
                    'Flag "non-existent" not found'
                );
            });
        });

        describe('stringVariationOrThrow', () => {
            it('should return value for enabled string flag', () => {
                expect(client.stringVariationOrThrow('string-variant')).toBe('hello world');
            });

            it('should throw for non-existent flag', () => {
                expect(() => client.stringVariationOrThrow('non-existent')).toThrow(
                    'Flag "non-existent" not found'
                );
            });

            it('should throw for disabled flag', () => {
                expect(() => client.stringVariationOrThrow('bool-disabled')).toThrow(
                    'Flag "bool-disabled" is disabled'
                );
            });
        });

        describe('numberVariationOrThrow', () => {
            it('should return value for valid number flag', () => {
                expect(client.numberVariationOrThrow('number-variant')).toBe(42.5);
            });

            it('should throw for invalid number payload', () => {
                expect(() => client.numberVariationOrThrow('invalid-number')).toThrow(
                    'Flag "invalid-number" has invalid number payload'
                );
            });
        });

        describe('jsonVariationOrThrow', () => {
            it('should return value for valid json flag', () => {
                expect(client.jsonVariationOrThrow('json-variant')).toEqual({
                    key: 'value',
                    nested: { a: 1 },
                });
            });

            it('should throw for invalid json payload', () => {
                expect(() => client.jsonVariationOrThrow('invalid-json')).toThrow(
                    'Flag "invalid-json" has invalid JSON payload'
                );
            });
        });
    });

    describe('getVariant', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return variant for existing flag', () => {
            const variant = client.getVariant('string-variant');

            expect(variant).toEqual({
                name: 'variantA',
                enabled: true,
                payload: 'hello world',
            });
        });

        it('should return fallback variant for non-existent flag', () => {
            const variant = client.getVariant('non-existent');

            expect(variant).toEqual({
                name: 'disabled',
                enabled: false,
            });
        });
    });

    describe('context management', () => {
        let consoleErrorSpy: jest.SpyInstance;

        beforeEach(async () => {
            await client.init();
            // Mock console.error to suppress expected error output during context changes
            consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => { });
            // Mock fetch to return empty flags response
            mockFetch.mockResolvedValue({
                ok: true,
                status: 200,
                json: async () => ({ success: true, data: { flags: bootstrapFlags } }),
            });
        });

        afterEach(() => {
            consoleErrorSpy.mockRestore();
        });

        it('should get initial context from config', () => {
            const configWithContext: GatrixClientConfig = {
                ...defaultConfig,
                context: { userId: 'user-123', deviceId: 'device-456' },
            };
            const clientWithContext = new FeaturesClient(emitter, configWithContext);

            expect(clientWithContext.getContext()).toMatchObject({
                userId: 'user-123',
                deviceId: 'device-456',
            });
        });

        it('should update context', () => {
            client.updateContext({ userId: 'new-user' });

            expect(client.getContext()).toMatchObject({ userId: 'new-user' });
        });

        it('should merge context with updateContext', () => {
            client.updateContext({ userId: 'user-1' });
            client.updateContext({ deviceId: 'device-1' });

            expect(client.getContext()).toMatchObject({
                userId: 'user-1',
                deviceId: 'device-1',
            });
        });

        it('should set individual context field', () => {
            client.setContextField('userId', 'user-123');

            expect(client.getContext().userId).toBe('user-123');
        });

        it('should remove context field', () => {
            client.updateContext({ userId: 'user-1', deviceId: 'device-1' });
            client.removeContextField('userId');

            expect(client.getContext()).toMatchObject({ deviceId: 'device-1' });
            expect(client.getContext().userId).toBeUndefined();
        });
    });

    describe('error handling', () => {
        it('should emit ERROR event on fetch failure', async () => {
            const errorCallback = jest.fn();
            emitter.on(EVENTS.ERROR, errorCallback);

            mockFetch.mockRejectedValueOnce(new Error('Network error'));

            const errorClient = new FeaturesClient(emitter, {
                ...defaultConfig,
                features: { disableRefresh: true },
            });

            await errorClient.init();
            // Try to fetch which will fail
            await errorClient.fetchFlags();

            expect(errorCallback).toHaveBeenCalled();
            expect(errorClient.getError()).toBeTruthy();
        });
    });

    describe('getAllFlags', () => {
        beforeEach(async () => {
            await client.init();
        });

        it('should return all flags', () => {
            const flags = client.getAllFlags();

            expect(flags).toHaveLength(bootstrapFlags.length);
            expect(flags.map((f) => f.name)).toEqual(expect.arrayContaining(bootstrapFlags.map((f) => f.name)));
        });
    });

    describe('offline mode', () => {
        it('should start successfully with bootstrap data in offline mode', async () => {
            const offlineClient = new FeaturesClient(emitter, {
                ...defaultConfig,
                offlineMode: true,
                features: {
                    bootstrap: bootstrapFlags,
                },
            });

            await offlineClient.init();
            await offlineClient.start();

            expect(offlineClient.isReady()).toBe(true);
            expect(offlineClient.isEnabled('bool-enabled')).toBe(true);
        });

        it('should throw error when offline mode has no data', async () => {
            const offlineClient = new FeaturesClient(emitter, {
                ...defaultConfig,
                offlineMode: true,
                features: {
                    bootstrap: [], // Empty bootstrap
                },
            });

            await offlineClient.init();

            await expect(offlineClient.start()).rejects.toThrow(
                'offlineMode requires bootstrap data or cached flags'
            );
        });

        it('should warn and skip fetchFlags in offline mode', async () => {
            const mockLogger = {
                debug: jest.fn(),
                info: jest.fn(),
                warn: jest.fn(),
                error: jest.fn(),
            };
            const offlineClient = new FeaturesClient(emitter, {
                ...defaultConfig,
                offlineMode: true,
                logger: mockLogger,
                features: {
                    bootstrap: bootstrapFlags,
                },
            });

            await offlineClient.init();
            await offlineClient.start();

            // Try to fetch - should warn and do nothing
            await offlineClient.fetchFlags();

            expect(mockLogger.warn).toHaveBeenCalledWith(
                expect.stringContaining('offline mode')
            );
        });

        it('should not start polling in offline mode', async () => {
            const offlineClient = new FeaturesClient(emitter, {
                ...defaultConfig,
                offlineMode: true,
                features: {
                    bootstrap: bootstrapFlags,
                    refreshInterval: 1, // 1 second
                },
            });

            await offlineClient.init();
            await offlineClient.start();

            // Mock should not be called even after waiting
            expect(mockFetch).not.toHaveBeenCalled();
        });
    });
});

File: packages/sdks/client-sdks/gatrix-flutter-sdk/test/features_client_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gatrix_flutter_sdk/src/models.dart';
import 'package:gatrix_flutter_sdk/src/events.dart';
import 'package:gatrix_flutter_sdk/src/features_client.dart';

// ==================== Test Helpers ====================

EvaluatedFlag makeFlag({
  required String name,
  bool enabled = true,
  String variantName = 'disabled',
  bool variantEnabled = false,
  dynamic value,
  ValueType valueType = ValueType.none,
  int version = 1,
  bool impressionData = false,
  String? reason,
}) {
  return EvaluatedFlag(
    name: name,
    enabled: enabled,
    variant: Variant(name: variantName, enabled: variantEnabled, value: value),
    valueType: valueType,
    version: version,
    impressionData: impressionData,
    reason: reason,
  );
}

FeaturesClient makeClient({
  List<EvaluatedFlag>? bootstrap,
  bool explicitSyncMode = false,
}) {
  final client = FeaturesClient(
    apiUrl: 'https://test.example.com',
    apiToken: 'test-token',
    appName: 'test-app',
    environment: 'test',
    context: GatrixContext(userId: 'test-user'),
    events: EventEmitter(),
    explicitSyncMode: explicitSyncMode,
  );
  if (bootstrap != null) {
    client.applyBootstrapFlags(bootstrap);
  }
  return client;
}

final List<EvaluatedFlag> fullBootstrap = [
  makeFlag(
    name: 'bool-flag',
    enabled: true,
    variantName: 'on',
    variantEnabled: true,
    value: true,
    valueType: ValueType.boolean,
    reason: 'default_variant',
  ),
  makeFlag(
    name: 'string-flag',
    enabled: true,
    variantName: 'greeting',
    variantEnabled: true,
    value: 'hello world',
    valueType: ValueType.string,
  ),
  makeFlag(
    name: 'number-flag',
    enabled: true,
    variantName: 'rate-limit',
    variantEnabled: true,
    value: 42,
    valueType: ValueType.number,
  ),
  makeFlag(
    name: 'float-flag',
    enabled: true,
    variantName: 'ratio',
    variantEnabled: true,
    value: 3.14,
    valueType: ValueType.number,
  ),
  makeFlag(
    name: 'json-flag',
    enabled: true,
    variantName: 'config',
    variantEnabled: true,
    value: {'key': 'value', 'nested': {'a': 1}},
    valueType: ValueType.json,
  ),
  makeFlag(
    name: 'disabled-flag',
    enabled: false,
    variantName: r'$disabled',
    variantEnabled: false,
    value: false,
    valueType: ValueType.boolean,
  ),
];

void main() {
  group('FeaturesClient - Flag Access', () {
    test('hasFlag returns true for existing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.hasFlag('bool-flag'), true);
      expect(client.hasFlag('nonexistent'), false);
    });

    test('getAllFlags returns all bootstrapped flags', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.getAllFlags().length, fullBootstrap.length);
    });

    test('getFlag returns FlagProxy for existing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('bool-flag');
      expect(proxy.exists, true);
      expect(proxy.name, 'bool-flag');
    });

    test('getFlag returns FlagProxy with exists=false for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('nonexistent');
      expect(proxy.exists, false);
    });
  });

  group('FeaturesClient - isEnabled', () {
    test('returns true for enabled flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.isEnabled('bool-flag'), true);
    });

    test('returns false for disabled flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.isEnabled('disabled-flag'), false);
    });

    test('returns false for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.isEnabled('nonexistent'), false);
    });
  });

  group('FeaturesClient - getVariant', () {
    test('returns variant for existing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final variant = client.getVariant('bool-flag');
      expect(variant.name, 'on');
      expect(variant.enabled, true);
    });

    test('returns missingVariant for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final variant = client.getVariant('nonexistent');
      expect(variant.name, r'$missing');
    });
  });

  group('FeaturesClient - variation()', () {
    test('returns variant name for existing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.variation('bool-flag', 'fallback'), 'on');
    });

    test('returns variant name regardless of enabled state', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.variation('disabled-flag', 'fallback'), r'$disabled');
    });

    test('returns missingValue for non-existent flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.variation('nonexistent', 'default'), 'default');
    });
  });

  group('FeaturesClient - boolVariation()', () {
    test('returns boolean value for boolean flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.boolVariation('bool-flag', false), true);
    });

    test('returns missingValue for non-boolean valueType', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.boolVariation('string-flag', true), true);
    });

    test('returns missingValue for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.boolVariation('nonexistent', true), true);
    });

    test('disabled flag with boolean valueType returns value', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.boolVariation('disabled-flag', true), false);
    });
  });

  group('FeaturesClient - stringVariation()', () {
    test('returns string value for string flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.stringVariation('string-flag', ''), 'hello world');
    });

    test('returns missingValue for non-string valueType', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.stringVariation('bool-flag', 'def'), 'def');
    });

    test('returns missingValue for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.stringVariation('nonexistent', 'def'), 'def');
    });
  });

  group('FeaturesClient - intVariation()', () {
    test('returns int value for number flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.intVariation('number-flag', 0), 42);
    });

    test('truncates double to int', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.intVariation('float-flag', 0), 3);
    });

    test('returns missingValue for non-number valueType', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.intVariation('bool-flag', 99), 99);
    });

    test('returns missingValue for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.intVariation('nonexistent', 7), 7);
    });
  });

  group('FeaturesClient - doubleVariation()', () {
    test('returns double value for number flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.doubleVariation('float-flag', 0.0), 3.14);
    });

    test('converts int to double', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.doubleVariation('number-flag', 0.0), 42.0);
    });

    test('returns missingValue for non-number valueType', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.doubleVariation('string-flag', 99.0), 99.0);
    });

    test('returns missingValue for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.doubleVariation('nonexistent', 7.0), 7.0);
    });
  });

  group('FeaturesClient - jsonVariation()', () {
    test('returns map value for json flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.jsonVariation<Map<String, dynamic>>(
          'json-flag', <String, dynamic>{});
      expect(result['key'], 'value');
      expect((result['nested'] as Map)['a'], 1);
    });

    test('returns missingValue for non-json valueType', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.jsonVariation<Map<String, dynamic>>(
          'string-flag', {'default': true});
      expect(result['default'], true);
    });

    test('returns missingValue for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.jsonVariation<Map<String, dynamic>>(
          'nonexistent', {'fallback': true});
      expect(result['fallback'], true);
    });
  });

  group('FeaturesClient - Variation Details', () {
    test('boolVariationDetails for existing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.boolVariationDetails('bool-flag', false);
      expect(result.value, true);
      expect(result.flagExists, true);
      expect(result.enabled, true);
      expect(result.reason, 'default_variant');
    });

    test('boolVariationDetails for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.boolVariationDetails('nonexistent', true);
      expect(result.value, true);
      expect(result.flagExists, false);
      expect(result.reason, 'flag_not_found');
    });

    test('boolVariationDetails type mismatch returns missingValue + reason', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.boolVariationDetails('string-flag', true);
      expect(result.value, true);
      expect(result.flagExists, true);
      expect(result.reason, contains('type_mismatch'));
    });

    test('stringVariationDetails', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.stringVariationDetails('string-flag', '');
      expect(result.value, 'hello world');
      expect(result.flagExists, true);
    });

    test('intVariationDetails', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.intVariationDetails('number-flag', 0);
      expect(result.value, 42);
      expect(result.flagExists, true);
    });

    test('doubleVariationDetails', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.doubleVariationDetails('float-flag', 0.0);
      expect(result.value, 3.14);
      expect(result.flagExists, true);
    });

    test('jsonVariationDetails', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client.jsonVariationDetails<Map<String, dynamic>>(
          'json-flag', <String, dynamic>{});
      expect(result.value['key'], 'value');
      expect(result.flagExists, true);
    });
  });

  group('FeaturesClient - OrThrow Methods', () {
    test('boolVariationOrThrow returns value', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.boolVariationOrThrow('bool-flag'), true);
    });

    test('boolVariationOrThrow throws for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(() => client.boolVariationOrThrow('nonexistent'),
          throwsA(isA<GatrixException>()));
    });

    test('boolVariationOrThrow throws for type mismatch', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(() => client.boolVariationOrThrow('string-flag'),
          throwsA(isA<GatrixException>()));
    });

    test('stringVariationOrThrow returns value', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.stringVariationOrThrow('string-flag'), 'hello world');
    });

    test('stringVariationOrThrow throws for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(() => client.stringVariationOrThrow('nonexistent'),
          throwsA(isA<GatrixException>()));
    });

    test('intVariationOrThrow returns value', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.intVariationOrThrow('number-flag'), 42);
    });

    test('intVariationOrThrow throws for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(() => client.intVariationOrThrow('nonexistent'),
          throwsA(isA<GatrixException>()));
    });

    test('doubleVariationOrThrow returns value', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(client.doubleVariationOrThrow('float-flag'), 3.14);
    });

    test('doubleVariationOrThrow throws for type mismatch', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(() => client.doubleVariationOrThrow('bool-flag'),
          throwsA(isA<GatrixException>()));
    });

    test('jsonVariationOrThrow returns value', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final result = client
          .jsonVariationOrThrow<Map<String, dynamic>>('json-flag');
      expect(result['key'], 'value');
    });

    test('jsonVariationOrThrow throws for missing flag', () {
      final client = makeClient(bootstrap: fullBootstrap);
      expect(
          () => client
              .jsonVariationOrThrow<Map<String, dynamic>>('nonexistent'),
          throwsA(isA<GatrixException>()));
    });
  });

  group('FeaturesClient - Explicit Sync', () {
    test('isExplicitSyncEnabled returns correct value', () {
      final client1 = makeClient(explicitSyncMode: false);
      expect(client1.isExplicitSyncEnabled(), false);

      final client2 = makeClient(explicitSyncMode: true);
      expect(client2.isExplicitSyncEnabled(), true);
    });

    test('hasPendingSyncFlags returns false when not in sync mode', () {
      final client = makeClient(
          bootstrap: fullBootstrap, explicitSyncMode: false);
      expect(client.hasPendingSyncFlags(), false);
    });

    test('hasPendingSyncFlags returns true when flags are pending', () {
      final client = makeClient(
          bootstrap: fullBootstrap, explicitSyncMode: true);
      expect(client.hasPendingSyncFlags(), true);
    });
  });

  group('FeaturesClient - FlagProxy Delegation', () {
    test('FlagProxy delegates boolVariation to client', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('bool-flag');
      expect(proxy.boolVariation(false), true);
    });

    test('FlagProxy delegates stringVariation to client', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('string-flag');
      expect(proxy.stringVariation(''), 'hello world');
    });

    test('FlagProxy delegates intVariation to client', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('number-flag');
      expect(proxy.intVariation(0), 42);
    });

    test('FlagProxy delegates doubleVariation to client', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('float-flag');
      expect(proxy.doubleVariation(0.0), 3.14);
    });

    test('FlagProxy delegates jsonVariation to client', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('json-flag');
      final result = proxy.jsonVariation<Map<String, dynamic>>({});
      expect(result['key'], 'value');
    });

    test('FlagProxy delegates boolVariationOrThrow to client', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('bool-flag');
      expect(proxy.boolVariationOrThrow(), true);
    });

    test('FlagProxy delegates missing flag correctly', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final proxy = client.getFlag('nonexistent');
      expect(proxy.boolVariation(true), true);
      expect(proxy.stringVariation('def'), 'def');
      expect(proxy.intVariation(99), 99);
      expect(proxy.doubleVariation(99.0), 99.0);
    });
  });

  group('FeaturesClient - Stats', () {
    test('getStats returns expected structure', () {
      final client = makeClient(bootstrap: fullBootstrap);
      final stats = client.getStats();
      expect(stats['totalFlagCount'], fullBootstrap.length);
      expect(stats.containsKey('fetchFlagsCount'), true);
      expect(stats.containsKey('errorCount'), true);
      expect(stats.containsKey('consecutiveFailures'), true);
    });
  });
}
